### PR TITLE
Refactor RPC and implement hello protocol

### LINKED ISSUE #194: [TASK]🗃️ Updates from Spec | Hello Protocol
**Task summary**

Last week Filecoin spec team updated the `Hello Protocol` under Nodes/Network. In particular, the protocol string and flow. See the below links for detailed changes.

**Specification reference**

- [Hello Protocol PR](https://github.com/filecoin-project/specs/pull/847/files)
- [Specification Section Updated](https://filecoin-project.github.io/specs/#hello-spec)

### PR DESCRIPTION
**Pull request by austinabell**
**Summary of changes**
Changes introduced in this pull request:
- Refactors RPC to allow for more protocols
  - I started doing a bigger refactor, but it was very extensive and didn't think it was worth the time since it works like this. Can revisit though once we update libp2p dependency
- Implements hello protocol
  - Needed to add another protocol to test so this is an alternative PR to #245 
- Updated RPC tests to more efficiently use futures and not waste CPU polling unnecessarily

The hello protocol matches the spec definition but does not match lotus, lotus has only one Hello message type and has optional map fields that they serialize for both the request and response. It was a huge pain to try to retrieve their exact format so I will leave this for now but will probably have to switch it to match lotus to test our syncing functionality.

**Reference issue to close (if applicable)**

Closes #194 

**Other information and links**

### CODE REVIEW

File: node/forest_libp2p/src/rpc/mod.rs [hunk 1]
```diff
@@ -51,3 +50,36 @@ impl std::fmt::Display for RPCEvent {
         }
     }
 }
+
```
Comment (dutterbutter) [inline node/forest_libp2p/src/rpc/mod.rs]: For consistency should this be its own file `error.rs`?

File: node/forest_libp2p/src/rpc/protocol.rs [hunk 2]
```diff
@@ -10,6 +12,13 @@ use libp2p::core::{Negotiated, UpgradeInfo};
...
 use std::pin::Pin;
 
+/// RPCResponse payloads for request/response calls
+#[derive(Debug, Clone, PartialEq)]
+pub enum RPCResponse {
+    Blocksync(BlockSyncResponse),
+    Hello(HelloResponse),
+}
+
 /// Protocol upgrade for inbound RPC requests. Currently supports Blocksync.
```
Comment (dutterbutter) [inline node/forest_libp2p/src/rpc/protocol.rs]: Comment should be updated to include mention of hello protocol?

File: node/forest_libp2p/src/rpc/protocol.rs [hunk 3]
```diff
@@ -35,34 +44,53 @@ where
...
     fn protocol_info(&self) -> Self::InfoIter {
-        vec![b"/fil/sync/blk/0.0.1"]
+        self.supported_protocols()
+    }
+}
+
+impl RPCRequest {
+    pub fn supported_protocols(&self) -> Vec<&'static [u8]> {
+        match self {
+            // add more protocols when versions/encodings are supported
```
Comment (dutterbutter) [inline node/forest_libp2p/src/rpc/protocol.rs]: maybe add a TODO here

File: node/forest_libp2p/src/service.rs [hunk 4]
```diff
@@ -145,15 +152,19 @@ impl Libp2pService {
...
+                                RPCEvent::Request(req_id, RPCRequest::Blocksync(r)) => {
+                                    // TODO implement handling incoming blocksync requests
                                     swarm_stream.get_mut().send_rpc(peer_id, RPCEvent::Response(1, RPCResponse::Blocksync(BlockSyncResponse {
                                         chain: vec![],
                                         status: 203,
                                         message: "handling requests not implemented".to_owned(),
                                     })));
                                 }
+                                RPCEvent::Request(req_id, RPCRequest::Hello(message)) => {
+
```
Comment (dutterbutter) [inline node/forest_libp2p/src/service.rs]: nit: remove spacing

**Comment by dutterbutter** [comment]
This should also close #194

**Comment by austinabell** [comment]
> This should also close #194

It has that in the PR description already :)

**Pull request merged**